superset-frontend/src/explore/components/controls/DatasourceControl/index.jsx

```diff
@@ -117,6 +118,15 @@ class DatasourceControl extends React.PureComponent {
 
   onDatasourceSave(datasource) {
     this.props.actions.setDatasource(datasource);
+    // remove time column in the form_data
+    const timeCol = this.props.form_data?.granularity_sqla; // eslint-disable-line camelcase
+    const { columns } = this.props.datasource;
+    if (
+      datasource.type === 'table' &&
+      !columns.find(({ column_name }) => column_name === timeCol)?.is_dttm // eslint-disable-line camelcase
+    ) {
+      this.props.actions.setControlValue('granularity_sqla', null);
+    }
```

Comment:
When the time column control is empty and a column is made `is_dttm`, the time column control stays empty, despite being required (not possible to remove value once set). Would it make sense to default to the first temporal column if the value is unset? Something like this:
```diff
diff --git a/superset-frontend/src/explore/components/controls/DatasourceControl/index.jsx b/superset-frontend/src/explore/components/controls/DatasourceControl/index.jsx
index 873067565..44452fb0c 100644
--- a/superset-frontend/src/explore/components/controls/DatasourceControl/index.jsx
+++ b/superset-frontend/src/explore/components/controls/DatasourceControl/index.jsx
@@ -121,12 +121,20 @@ class DatasourceControl extends React.PureComponent {
     // remove time column in the form_data
     const timeCol = this.props.form_data?.granularity_sqla; // eslint-disable-line camelcase
     const { columns } = this.props.datasource;
+    const firstDttmCol = columns.find(column => column.is_dttm);
     if (
       datasource.type === 'table' &&
+      timeCol &&
       !columns.find(({ column_name }) => column_name === timeCol)?.is_dttm // eslint-disable-line camelcase
     ) {
       this.props.actions.setControlValue('granularity_sqla', null);
+    } else if (datasource.type === 'table' && !timeCol && firstDttmCol) {
+      this.props.actions.setControlValue(
+        'granularity_sqla',
+        firstDttmCol.column_name,
+      );
     }
+
     if (this.props.onDatasourceSave) {
       this.props.onDatasourceSave(datasource);
     }
```

Reply:
thanks for your comments!  This has the side effect of leaving an unexpected column to do the filtering and grain(the other words, first datetime column depends on the order of the columns in the database.). What do you think about?

Reply:
Done this change. Thanks, comments!